src/libopensc/card-piv.c

```diff
@@ -287,242 +546,1806 @@ struct piv_object {
+
+	/* use for several hash operations */
+	if ((hash_ctx = EVP_MD_CTX_new()) == NULL) {
+		r = SC_ERROR_OUT_OF_MEMORY;
+		goto err;
+	}
+
+	/* Step 1 set CBh = 0 */
+	CBh = 0;
+
+	/* Step H2 generate ephemeral EC */
+
+	if ((eph_ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_EC, NULL)) == NULL
+			|| EVP_PKEY_keygen_init(eph_ctx) <= 0
+			|| EVP_PKEY_CTX_set_ec_paramgen_curve_nid(eph_ctx, cs->nid) <= 0
+			|| EVP_PKEY_keygen(eph_ctx, &eph_pkey) <= 0
+			|| (eph_eckey = EVP_PKEY_get0_EC_KEY(eph_pkey)) == NULL
+			|| (eph_group = EC_KEY_get0_group(eph_eckey)) == NULL
+			|| (Qehlen = EC_POINT_point2oct(eph_group, EC_KEY_get0_public_key(eph_eckey),
+				POINT_CONVERSION_UNCOMPRESSED, NULL, Qehlen, NULL)) <= 0 /* get length */
+			|| Qehlen > cs->Qlen
+			|| (Qehlen = EC_POINT_point2oct(eph_group, EC_KEY_get0_public_key(eph_eckey),
+				POINT_CONVERSION_UNCOMPRESSED, Qeh, Qehlen, NULL)) <= 0
+			|| Qehlen > cs->Qlen) {
+		sc_log(card->ctx,"OpenSSL failed to create ephemeral EC key");
+		piv_log_openssl(card->ctx);
+		r = SC_ERROR_SM_AUTHENTICATION_FAILED;
+		goto err;
+	}
+	
+	/* For later use, get  Qeh without 04 and full size  X || Y */
+	if (Q2OS(cs->field_length, Qeh, Qehlen, Qeh_OS, &Qeh_OSlen)) {
+		sc_log(card->ctx,"Q2OS for Qeh failed");
+		r = SC_ERROR_INTERNAL;
+		goto err;
+	}
+
+	len2a = sc_asn1_put_tag(0x81, NULL, 1 + cs->IDshlen + Qehlen, NULL, 0, NULL);
+	len2b = sc_asn1_put_tag(0x80, NULL, 0, NULL, 0, NULL);
+	sbuflen = sc_asn1_put_tag(0x7C, NULL, len2a + len2b, NULL, 0, NULL);
+	if (sbuflen <= 0) {
+		r = SC_ERROR_INTERNAL;
+		goto err;
+	}
+	sbuf = malloc(sbuflen);
+	if (sbuf == NULL) {
+		r = SC_ERROR_OUT_OF_MEMORY;
+		goto err;
+	}
+	p = sbuf;
+
+	r =  sc_asn1_put_tag(0x7C, NULL, len2a + len2b, sbuf, sbuflen, &p);
 	if (r != SC_SUCCESS)
-		LOG_FUNC_RETURN(card->ctx, r);
+		goto err;
 
-	sc_format_apdu(card, &apdu,
-			recvbuf ? SC_APDU_CASE_4_SHORT: SC_APDU_CASE_3_SHORT,
-			ins, p1, p2);
-	apdu.flags |= SC_APDU_FLAGS_CHAINING;
-	apdu.lc = sendbuflen;
-	apdu.datalen = sendbuflen;
-	apdu.data = sendbuf;
+	r = sc_asn1_put_tag(0x81, NULL, 1 + cs->IDshlen + Qehlen, p, sbuflen - (p - sbuf), &p);
+	if (r != SC_SUCCESS)
+		goto err;
 
-	if (recvbuf && recvbuflen) {
-		apdu.le = (recvbuflen > 256) ? 256 : recvbuflen;
-		apdu.resplen = recvbuflen;
-	} else {
-		 apdu.le = 0;
-		 apdu.resplen = 0;
+	/* Step H1 set CBh to 0x00 */
+	*p++ = CBh;
+
+#ifdef WIN32
+	pid = (unsigned long) GetCurrentProcessId();
+#else
+	pid = (unsigned long) getpid(); /* use PID as our ID so different from other processes */
+#endif
+	memcpy(IDsh, &pid, MIN(sizeof(pid), cs->IDshlen));
+	memcpy(p, IDsh, cs->IDshlen);
+	p += cs->IDshlen;
+	memcpy(p, Qeh, Qehlen);
+	p += Qehlen;
+	
+	r = sc_asn1_put_tag(0x82, NULL, 0, p, sbuflen - (p - sbuf), &p); /* null data */
+	if (r != SC_SUCCESS)
+		goto err;
+
+	/* Step H3 send CBh||IDsh|| Qeh  Qeh in 04||x||y */
+	/* Or call sc_transmit directly */
+	r = piv_general_io(card, 0x87, cs->p1, 0x04, sbuf, (p - sbuf), rbuf, rbuflen);
```

Comment:
If one was going to check the code against the authoritative source it would by NIST 800-73-4, Part 2, Section 4.1.8. (see below). It uses '87' as the commend, (which happens to be GENERAL AUTHENTICATE)  They don't use defines and using defines in OpenSC just adds complexity. I still don't want to change it.  

4.1.8 Command Interface
The following command interface shall be used for the key establishment protocol.
Command Syntax

CLA    '00'

INS     '87'

P1      Algorithm reference ('27' or '2E'), as specified in the 0xAC tag of the application property template

P2     '04' (PIV Secure Messaging key).

Lc      Length of data field

Data Field     '7C' L1 { '81' L2 { CBH || IDsH || QeH } '82 00' }, where CBH is 0x00, IDsH is an 8-byte client application identifier as described in Section 4.1.3, and QeH is an ephemeral public key encoded as 04 || X || Y, as specified in the “Value” column of Table 13.

Le '00'